### PR TITLE
[pkg/stanza] when the 'jsonArrayParserFeatureGate'  is disabled, the func 'Build'  returns an error

### PR DESCRIPTION
**Pull request by ahern-pro**
**Description:** <Describe what has changed.>

bug: 
- collector launch fail with '--feature-gates=logs.jsonParserArray'

how to fix:
1. `func init()` always register operator to `operator.DefaultRegistry`.
2. when the `jsonArrayParserFeatureGate` is disabled, the func `Build` returns an error.

**Link to tracking Issue:** <Issue number if applicable>
https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/32313

### CODE REVIEW

File: .chloggen/fix-jsonarrayparser-build.yaml [hunk 1]
```diff
@@ -0,0 +1,27 @@
+# Use this changelog template to create an entry for release notes.
+
+# One of 'breaking', 'deprecation', 'new_component', 'enhancement', 'bug_fix'
+change_type: bug_fix
+
+# The name of the component, or a single word describing the area of concern, (e.g. filelogreceiver)
+component: pkg/stanza
+
+# A brief description of the change.  Surround your text with quotes ("") if it needs to start with a backtick (`).
+note: "When the `jsonArrayParserFeatureGate` is disabled, the func `Build` returns an error."
```
Comment (djaglowski) [inline .chloggen/fix-jsonarrayparser-build.yaml:10]:
```suggestion
note: Fix race condition which prevented `jsonArrayParserFeatureGate` from working correctly.
```

File: pkg/stanza/operator/parser/jsonarray/parser_test.go [hunk 2]
```diff
@@ -8,23 +8,30 @@ import (
...
+	"go.opentelemetry.io/collector/featuregate"
 
 	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza/entry"
 	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza/operator"
 	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza/testutil"
 )
 
 func newTestParser(t *testing.T) *Parser {
+	err := featuregate.GlobalRegistry().Set("logs.jsonParserArray", true)
+	require.Nil(t, err)
```
Comment (djaglowski) [inline pkg/stanza/operator/parser/jsonarray/parser_test.go:20]: Let's use [`testutil.SetFeatureGateForTest`](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/29d1f4fb97d42f5ef09a3132570d64321a9c1449/internal/common/testutil/testutil.go#L130).

Can we also get the string value from `jsonArrayParserFeatureGate` or make it a constant?

```suggestion
	defer testutil.SetFeatureGateForTest(t, enableParser, true)
```

**Review by djaglowski — COMMENTED** [review]
Ok, the functionality looks correct to me now. Just a few nits for changelog and tests.

File: pkg/stanza/operator/parser/jsonarray/parser_test.go (hunk 2 above)
Comment (li-zeyuan) [reply pkg/stanza/operator/parser/jsonarray/parser_test.go:20]: Thanks for your suggestion. 
`SetFeatureGateForTest` resets the original state very carefully

**Comment by li-zeyuan** [comment]
build-and-test / coverage failed. Looks like it's GitHub rate limit.
```go
error - 2024-04-18 14:44:56,662 -- Commit creating failed: {"detail":"Tokenless has reached GitHub rate limit. Please upload using a token: https://docs.codecov.com/docs/adding-the-codecov-token. Expected available in 1017 seconds."}
Error: Codecov: Failed to properly create commit: The process '/home/runner/work/_actions/Wandalen/codecov-action/dist/codecov' failed with exit code 1
Error: Process returned exit code 1
```

**Review by djaglowski — APPROVED** [review]

**Pull request merged**